README.md
# CoverPop

CoverPop is modal splash page plugin with cookie integration. Style agnostic. Responsive friendly.

## Website

[View Website](http://coverpopjs.com)

## Demo

[View Demo](http://tylerpearson.github.io/CoverPop/)

## Usage

If you are trying to show up a lightbox popup similar first time visitors that is easily styleable and has cookie integration, this is for you.

If you are trying to show a colorbox image when someone clicks on an internal link, this isn't for you.

## Background

CoverPop is an extension of code that [we](http://newmediacampaigns.com "New Media Campaigns") frequently found ourselves using on political campaign websites for email signups with first time visitors. We needed something that was customizable to the point we could use it exactly how we wanted and didn't get in the way.

## Questions and feature requests

Please add an issue instead of sending me an email. It makes it easier to keep track and allows other people to participate in the discussion.

## Setup

### CSS

Include the plugin css file:

```html
<link rel="stylesheet" href="css/CoverPop.css">
```

### JS

CoverPop does not use jQuery.

Include the CoverPop.js file at the bottom of your markup.

```html
<script src="js/CoverPop.js"></script>
```

Start it with the default settings:

```javascript
CoverPop.start();
```

**CoverPop** can be customized:

```javascript
CoverPop.start({
    coverId:             'CoverPop-cover',       // set default cover id
    expires:             30,                     // duration (in days) before it pops up again
    closeClassNoDefault: 'CoverPop-close',       // close if someone clicks an element with this class and prevent default action
    closeClassDefault:   'CoverPop-close-go',    // close if someone clicks an element with this class and continue default action
    cookieName:          '_CoverPop',            // to change the plugin cookie name
    onPopUpOpen:         function() {},          // on popup open callback function
    onPopUpClose:        function() {},          // on popup close callback function
    forceHash:           'splash',               // hash to append to url to force display of popup (e.g. http://yourdomain.com/#splash)
    delayHash:           'go',                   // hash to append to url to delay popup for 1 day (e.g. http://yourdomain.com/#go)
    closeOnEscape:       true                    // close if the user clicks escape
    delay:               0                       // set an optional delay (in milliseconds) before showing the popup
});
```

### HTML

`#CoverPop-cover` is used on the full window cover.

By default, a click on any element with `.CoverPop-close` will close the popup. The plugin uses `preventDefault` with elements with this class.

If you wish to continue with the default action, but also hide the popup, add `.CoverPop-close-go`. This is particularly useful for form submissions that are sent to another page.

```html
<body>


  <!-- your site's markup -->


  <!-- start popup -->
  <div id="CoverPop-cover" class="splash">
      <div class="CoverPop-content splash-center">

          <!-- the popup content (form, welcome message, etc.) -->

          <a class="CoverPop-close" href="#">or skip signup</a>

      </div><!--end .splash-center -->
  </div><!--end .splash -->
  <!-- end popup -->

  <!-- js, etc. -->

</body>
```

## Examples

See all example uses at [www.coverpopjs.com](http://coverpopjs.com).

### Use a new cookie

*Example use:* You changed the popup, and want all new visitors to see the new one.

```js
CoverPop.start({
    cookieName: 'CoverPop-new'
});
```

### Delay showing the popup for 4 seconds

*Example use:* You'd prefer to have the popup not show immediately, and instead wait 4 seconds.

```
CoverPop.init({
    delay: 4000
});
```

### Hide popup to the visitor for a day

*Example use:* You are sending visitors to a page that already has a signup form, and don't want them to see the popup.

```js
http://www.example.com/#go
```

### Force popup

*Example use:* You are sending visitors to a page and want visitors who have already been to the site to see the popup again.

```js
http://www.example.com/#splash
```

### Change time before cookie expires

*Example use:* You want visitors to see the popup every 7 days, instead of 30.

```js
CoverPop.start({
    expires: 7
});
```

### Do something after the popup closes

*Example use:* You want to start a video after the popup is closed.

```js
CoverPop.start({
    onPopUpClose: function() {
        var player = document.getElementById('myVideo');
        player.play();
    }
});
```

### Hide the popup when someone clicks submit on a form and continue

*Example use:* You want to close the popup and set the cookie when a visitor clicks submit on a form in the popup.

```html
<input type="submit" value="Submit" class="CoverPop-close-go">
```


## License

MIT


CHANGELOG.md
## Updates

* v2.4.1 - Prevent child elements from inherting parent's events
* v2.4.0 - Added the ability to delay showing the popup
* v2.3.0 - Remove console.log debugging option
* v2.2.1 - Add fix for open callback check
* v2.2 - Add check to ensure cookies are enabled
* v2.0.2 - Tweak default CSS to allow scrolling on overflow
* v2.0.1 - Add underscore to cookie name to allow Varnish caching *7/27/2013*
* v2.1 *6/30/2013*
  * Ditch jQuery dependency
  * Switch to using an id instead of class for CoverPop-cover
  * Remove fadeIn/Out duration options because animations are now handled through CSS
  * Add CoverPop.start() alias for CoverPop.init()
* v2.0.0 - Major changes  *6/22/2013*
  * Major cleanup of the JS.
  * Removed the option to center with JS.
  * Namespace has been capitalized to CoverPop
  * Add [CoverPop site](http://coverpopjs.com) for better docs
* v1.0.4 - Add ability to delay popup by appending a hash to the url *6/13/2013*
* v1.0.3 - Add ability to set a class that still sets a cookie but continues with the default action (useful for submit buttons) *5/29/2013*
* v1.0.2 - Add option to close popup by hitting escape *5/23/2013*
* v1.0.1 - Add ability to force the popup by adding a hash to the url *5/22/2013*

js/jquery.coverPop.js
// coverPop v1.0.4 - jQuery Plugin
// License: http://www.opensource.org/licenses/mit-license.php
// To use: $(document).coverPop()
// Will set up a full page popup cover overlay and hide for a set period of time

(function($) {

    'use strict';

    $.fn.coverPop = function(options) {

        // set default plugin settings
        var settings = {
            coverClass:          'coverPop-cover'       // set default cover class
        ,   contentClass:        'coverPop-content'     // set default content class
        ,   fadeInDuration:      500                    // time (in milliseconds) to fade in
        ,   fadeOutDuration:     500                    // time (in milliseconds) to fade out
        ,   expires:             30                     // duration (in days) before it pops up again
        ,   jsCenter:            false                  // if we want the plugin to center the middle box with js (nasty and unrecommended)
        ,   closeClassNoDefault: "coverPop-close"       // close if someone clicks an element with this class and prevent default action
        ,   closeClassDefault:   "coverPop-close-go"    // close if someone clicks an element with this class and continue default action
        ,   cookieName:          "coverPop"             // to change the plugin cookie name
        ,   onPopUpOpen:         function() {}          // on popup open / default is nothing
        ,   onPopUpClose:        function() {}          // on popup close / default is nothing
        ,   forceHash:           'splash'               // hash to append to url to force display of popup
        ,   delayHash:           'go'                   // hash to append to url to delay popup for 1 day
        ,   closeOnEscape:       true                   // close if the user clicks escape
        ,   info:                false                  // toggle console.log statements
        };

        $.extend(settings, options);

        // cache elements we will use
        var html    = $('html')
        ,   body    = $('body')
        ,   cover   = $('.' + settings.coverClass)
        ,   content = $('.' + settings.contentClass)


        return this.each(function() {

            var self = $(this);

            // js centering of element (if not set in css)
            function centerize(element){

                var contentHeight  = element.height()
                ,   contentWidth   = element.width()
                ,   windowWidth    = $(window).width()
                ,   windowHeight   = $(window).height();

                shareInfo("content height: " + contentHeight + ", content width: " + contentWidth);

                element.css({
                    "position" : "absolute",
                    "left"     : Math.round(windowWidth  / 2 - contentWidth / 2),
                    "top"      : Math.round((windowHeight / 2 - contentHeight / 2) - ((windowHeight / 100) * 5)),
                    "margin"   : 0
                });

                shareInfo("centered");
            }

            // open cover popup
            function openPopUp() {

                if (hashExists(delayHash)) {
                    // set the cookie for a day
                    setCookie(1);
                    return;
                }

                // when the popup opens
                function openCallback() {

                    // if there is a function callback on open
                    if (isFunction(settings.onPopUpOpen)) {
                        settings.onPopUpOpen.call(self);

                        shareInfo("CoverPop is open.");  // say it's open
                    }

                    // if js is centering
                    if (settings.jsCenter) {
                        // initial centering
                        centerize(settings.content);

                        // if the window moves centering
                        $(window).on('resize', function(){
                            centerize(settings.content);
                        });
                    }

                }


                // add open class to html
                html.addClass("coverPop-open");
                cover.fadeIn(settings.fadeInDuration, openCallback);

                // bind close events and prevent default event
                if (settings.closeClassNoDefault.length) {
                    $(document).on('click', '.' + settings.closeClassNoDefault, function(e) {
                        e.preventDefault();
                        closePopUp();
                    });
                }

                // bind close events and continue with default event
                if (settings.closeClassDefault.length) {
                    $(document).on('click', '.' + settings.closeClassDefault, function(e) {
                        closePopUp();
                    });
                }

                // bind escape detection to document
                $(document).bind('keyup', onDocUp);
            }

            // close the cover popup
            function closePopUp() {

                // add open class to html
                html.removeClass("coverPop-open");

                // when the popup closes
                function closeCallback() {
                    // if there is a function callback on close
                    if (isFunction(settings.onPopUpClose)) {
                        settings.onPopUpClose.call(self);

                        shareInfo("CoverPop is closed.");  // say it's closed
                    }
                }

                // add open class to html
                html.removeClass("coverPop-open");
                cover.fadeOut(settings.fadeOutDuration, closeCallback);

                // set the cookie so we don't keep seeing the popup
                setCookie(settings.expires);

                // unbind escape detection to document
                $(document).unbind('keyup', onDocUp);
            }

            // check if there is a cookie or hash before proceeding
            if (checkCookie() === false || hashExists(settings.forceHash) === true) {
                openPopUp();
            }

            // close popup when user hits escape button
            function onDocUp(e) {
                if (settings.closeOnEscape) {
                    if (e.keyCode === 27) {
                        closePopUp();
                    }
                }
            }


        });


        // test if it's a function
        function isFunction(functionToCheck) {
            var getType = {};
            return functionToCheck && getType.toString.call(functionToCheck) === '[object Function]';

        }


        // for info and debugging
        function shareInfo(message) {
            if (window.console && window.console.log && settings.info) {
                window.console.log(message);
            }
        }


        // function for setting cookie
        function setCookie(days) {
            var date = new Date();

            // get milliseconds at current time plus number of days
            date.setTime(+ date + (days * 86400000)); //24 * 60 * 60 * 1000
            document.cookie = settings.cookieName + "=true; expires=" + date.toGMTString() + "; path=/";

            shareInfo("Cookie " + settings.cookieName + " set for " + days + " days away.");
        }


        // check cookie exists and isn't expired
        function checkCookie() {
            // check that the cookie exists
            if (document.cookie.indexOf(settings.cookieName) !== -1) return true;

            // if the cookie doesn't exist
            return false;
        }


        // check if there is a hash in the url
        function hashExists(hash) {
            // check for hash
            if (window.location.hash.indexOf(hash) !== -1) return true;

            // if there isn't a hash
            return false;
        }

    };

}(jQuery));

js/v1/jquery.coverPop.v1.js
// coverPop v1.0.4 - jQuery Plugin
// License: http://www.opensource.org/licenses/mit-license.php
// To use: $(document).coverPop()
// Will set up a full page popup cover overlay and hide for a set period of time

(function($) {

    'use strict';

    $.fn.coverPop = function(options) {

        // set default plugin settings
        var settings = {
            coverClass:          'coverPop-cover'       // set default cover class
        ,   contentClass:        'coverPop-content'     // set default content class
        ,   fadeInDuration:      500                    // time (in milliseconds) to fade in
        ,   fadeOutDuration:     500                    // time (in milliseconds) to fade out
        ,   expires:             30                     // duration (in days) before it pops up again
        ,   jsCenter:            false                  // if we want the plugin to center the middle box with js (nasty and unrecommended)
        ,   closeClassNoDefault: "coverPop-close"       // close if someone clicks an element with this class and prevent default action
        ,   closeClassDefault:   "coverPop-close-go"    // close if someone clicks an element with this class and continue default action
        ,   cookieName:          "coverPop"             // to change the plugin cookie name
        ,   onPopUpOpen:         function() {}          // on popup open / default is nothing
        ,   onPopUpClose:        function() {}          // on popup close / default is nothing
        ,   forceHash:           'splash'               // hash to append to url to force display of popup
        ,   delayHash:           'go'                   // hash to append to url to delay popup for 1 day
        ,   closeOnEscape:       true                   // close if the user clicks escape
        ,   info:                false                  // toggle console.log statements
        };

        $.extend(settings, options);

        // cache elements we will use
        var html    = $('html')
        ,   body    = $('body')
        ,   cover   = $('.' + settings.coverClass)
        ,   content = $('.' + settings.contentClass);


        return this.each(function() {

            var self = $(this);

            // js centering of element (if not set in css)
            function centerize(element){

                var contentHeight  = element.height()
                ,   contentWidth   = element.width()
                ,   windowWidth    = $(window).width()
                ,   windowHeight   = $(window).height();

                shareInfo("content height: " + contentHeight + ", content width: " + contentWidth);

                element.css({
                    "position" : "absolute",
                    "left"     : Math.round(windowWidth  / 2 - contentWidth / 2),
                    "top"      : Math.round((windowHeight / 2 - contentHeight / 2) - ((windowHeight / 100) * 5)),
                    "margin"   : 0
                });

                shareInfo("centered");
            }

            // open cover popup
            function openPopUp() {

                if (hashExists(delayHash)) {
                    // set the cookie for a day
                    setCookie(1);
                    return;
                }

                // when the popup opens
                function openCallback() {

                    // if there is a function callback on open
                    if (isFunction(settings.onPopUpOpen)) {
                        settings.onPopUpOpen.call(self);
                        shareInfo("CoverPop is open.");  // say it's open
                    }

                    if (settings.jsCenter) {
                        centerize(settings.content);
                        $(window).on('resize', function(){
                            centerize(settings.content);
                        });
                    }

                }


                html.addClass("coverPop-open");
                cover.fadeIn(settings.fadeInDuration, openCallback);

                // bind close events and prevent default event
                if (settings.closeClassNoDefault.length) {
                    $(document).on('click', '.' + settings.closeClassNoDefault, function(e) {
                        e.preventDefault();
                        closePopUp();
                    });
                }

                // bind close events and continue with default event
                if (settings.closeClassDefault.length) {
                    $(document).on('click', '.' + settings.closeClassDefault, function(e) {
                        closePopUp();
                    });
                }

                // bind escape detection to document
                $(document).bind('keyup', onDocUp);
            }

            // close the cover popup
            function closePopUp() {

                html.removeClass("coverPop-open");

                // when the popup closes
                function closeCallback() {
                    // if there is a function callback on close
                    if (isFunction(settings.onPopUpClose)) {
                        settings.onPopUpClose.call(self);

                        shareInfo("CoverPop is closed.");  // say it's closed
                    }
                }

                html.removeClass("coverPop-open");
                cover.fadeOut(settings.fadeOutDuration, closeCallback);

                // set the cookie so we don't keep seeing the popup
                setCookie(settings.expires);

                // unbind escape detection to document
                $(document).unbind('keyup', onDocUp);
            }

            // check if there is a cookie or hash before proceeding
            if (checkCookie() === false || hashExists(settings.forceHash) === true) {
                openPopUp();
            }

            // close popup when user hits escape button
            function onDocUp(e) {
                if (settings.closeOnEscape) {
                    if (e.keyCode === 27) {
                        closePopUp();
                    }
                }
            }


        });


        // test if it's a function
        function isFunction(functionToCheck) {
            var getType = {};
            return functionToCheck && getType.toString.call(functionToCheck) === '[object Function]';
        }


        // for info and debugging
        function shareInfo(message) {
            if (window.console && window.console.log && settings.info) {
                window.console.log(message);
            }
        }


        // for setting cookie
        function setCookie(days) {
            var date = new Date();
            date.setTime(+ date + (days * 86400000)); //24 * 60 * 60 * 1000
            document.cookie = settings.cookieName + "=true; expires=" + date.toGMTString() + "; path=/";
            shareInfo("Cookie " + settings.cookieName + " set for " + days + " days away.");
        }


        // check cookie exists and isn't expired
        function checkCookie() {
            if (document.cookie.indexOf(settings.cookieName) !== -1) return true;
            return false;
        }


        // check if there is a hash in the url
        function hashExists(hash) {
            if (window.location.hash.indexOf(hash) !== -1) return true;
            return false;
        }

    };

}(jQuery));

CoverPop.js
/*!
 * CoverPop 2.4.1
 * http://coverpopjs.com
 *
 * Copyright (c) 2014 Tyler Pearson
 * Licensed under the MIT license: http://www.opensource.org/licenses/mit-license.php
 */

(function (CoverPop, undefined) {

    'use strict';

    // set default settings
    var settings = {

            // set default cover id
            coverId: 'CoverPop-cover',

            // duration (in days) before it pops up again
            expires: 30,

            // close if someone clicks an element with this class and prevent default action
            closeClassNoDefault: 'CoverPop-close',

            // close if someone clicks an element with this class and continue default action
            closeClassDefault: 'CoverPop-close-go',

            // change the cookie name
            cookieName: '_CoverPop',

            // on popup open function callback
            onPopUpOpen: null,

            // on popup close function callback
            onPopUpClose: null,

            // hash to append to url to force display of popup
            forceHash: 'splash',

            // hash to append to url to delay popup for 1 day
            delayHash: 'go',

            // close if the user clicks escape
            closeOnEscape: true,

            // set an optional delay (in milliseconds) before showing the popup
            delay: 0
        },


        // grab the elements to be used
        $el = {
            html: document.getElementsByTagName('html')[0],
            cover: document.getElementById(settings.coverId),
            closeClassDefaultEls: document.querySelectorAll('.' + settings.closeClassDefault),
            closeClassNoDefaultEls: document.querySelectorAll('.' + settings.closeClassNoDefault)
        },


        /**
         * Helper methods
         */

        util = {

            hasClass: function(el, name) {
                return new RegExp('(\\s|^)' + name + '(\\s|$)').test(el.className);
            },

            addClass: function(el, name) {
                if (!util.hasClass(el, name)) {
                    el.className += (el.className ? ' ' : '') + name;
                }
            },

            removeClass: function(el, name) {
                if (util.hasClass(el, name)) {
                    el.className = el.className.replace(new RegExp('(\\s|^)' + name + '(\\s|$)'), ' ').replace(/^\s+|\s+$/g, '');
                }
            },

            addListener: function(target, type, handler) {
                if (target.addEventListener) {
                    target.addEventListener(type, handler, false);
                } else if (target.attachEvent) {
                    target.attachEvent('on' + type, handler);
                }
            },

            removeListener: function(target, type, handler) {
                if (target.removeEventListener) {
                    target.removeEventListener(type, handler, false);
                } else if (target.detachEvent) {
                    target.detachEvent('on' + type, handler);
                }
            },

            isFunction: function(functionToCheck) {
                var getType = {};
                return functionToCheck && getType.toString.call(functionToCheck) === '[object Function]';
            },

            setCookie: function(name, days) {
                var date = new Date();
                date.setTime(+ date + (days * 86400000));
                document.cookie = name + '=true; expires=' + date.toGMTString() + '; path=/';
            },

            hasCookie: function(name) {
                if (document.cookie.indexOf(name) !== -1) {
                    return true;
                }
                return false;
            },

            // check if there is a hash in the url
            hashExists: function(hash) {
                if (window.location.hash.indexOf(hash) !== -1) {
                    return true;
                }
                return false;
            },

            preventDefault: function(event) {
                if (event.preventDefault) {
                    event.preventDefault();
                } else {
                    event.returnValue = false;
                }
            },

            mergeObj: function(obj1, obj2) {
                for (var attr in obj2) {
                    obj1[attr] = obj2[attr];
                }
            }
        },


        /**
         * Private Methods
         */

        // close popup when user hits escape button
        onDocUp = function(e) {
            if (settings.closeOnEscape) {
                if (e.keyCode === 27) {
                    CoverPop.close();
                }
            }
        },

        openCallback = function() {

            // if not the default setting
            if (settings.onPopUpOpen !== null) {

                // make sure the callback is a function
                if (util.isFunction(settings.onPopUpOpen)) {
                    settings.onPopUpOpen.call();
                } else {
                    throw new TypeError('CoverPop open callback must be a function.');
                }
            }
        },

        closeCallback = function() {

            // if not the default setting
            if (settings.onPopUpClose !== null) {

                // make sure the callback is a function
                if (util.isFunction(settings.onPopUpClose)) {
                    settings.onPopUpClose.call();
                } else {
                    throw new TypeError('CoverPop close callback must be a function.');
                }
            }
        };



    /**
     * Public methods
     */

    CoverPop.open = function() {

        var i, len;

        if (util.hashExists(settings.delayHash)) {
            util.setCookie(settings.cookieName, 1); // expire after 1 day
            return;
        }

        util.addClass($el.html, 'CoverPop-open');

        // bind close events and prevent default event
        if ($el.closeClassNoDefaultEls.length > 0) {
            for (i=0, len = $el.closeClassNoDefaultEls.length; i < len; i++) {
                util.addListener($el.closeClassNoDefaultEls[i], 'click', function(e) {
                    if (e.target === this) {
                        util.preventDefault(e);
                        CoverPop.close();
                    }
                });
            }
        }

        // bind close events and continue with default event
        if ($el.closeClassDefaultEls.length > 0) {
            for (i=0, len = $el.closeClassDefaultEls.length; i < len; i++) {
                util.addListener($el.closeClassDefaultEls[i], 'click', function(e) {
                    if (e.target === this) {
                        CoverPop.close();
                    }
                });
            }
        }

        // bind escape detection to document
        util.addListener(document, 'keyup', onDocUp);
        openCallback();
    };

    CoverPop.close = function(e) {
        util.removeClass($el.html, 'CoverPop-open');
        util.setCookie(settings.cookieName, settings.expires);

        // unbind escape detection to document
        util.removeListener(document, 'keyup', onDocUp);
        closeCallback();
    };

    CoverPop.init = function(options) {
        if (navigator.cookieEnabled) {
            util.mergeObj(settings, options);

            // check if there is a cookie or hash before proceeding
            if (!util.hasCookie(settings.cookieName) || util.hashExists(settings.forceHash)) {
                if (settings.delay === 0) {
                    CoverPop.open();
                } else {
                    // delay showing the popup
                    setTimeout(CoverPop.open, settings.delay);
                }

            }
        }
    };

    // alias
    CoverPop.start = function(options) {
        CoverPop.init(options);
    };


}(window.CoverPop = window.CoverPop || {}));
